src/object/cli_obj.c

```diff
@@ -1230,12 +1246,102 @@ obj_req_tgts_dump(struct obj_req_tgts *req_tgts)
 	}
 }
 
+static int
+obj_ec_check_exist_tgts_get(struct dc_object *obj, uint32_t map_ver, uint32_t start_shard,
+			    uint32_t shard_cnt, struct obj_auxi_args *obj_auxi)
+{
+	struct obj_req_tgts	*req_tgts = &obj_auxi->req_tgts;
+	struct daos_oclass_attr	*oca;
+	struct daos_shard_tgt	*tgt = NULL;
+	uint32_t		 psize, dsize, shard, i;
+	int			 rc = 0;
+
+	/* TODO  need change after parity shard rotate */
+	oca = obj_get_oca(obj);
+	psize = obj_ec_parity_tgt_nr(oca);
+	dsize = obj_ec_data_tgt_nr(oca);
+	D_ASSERT(start_shard % (psize + dsize) == 0);
+	D_ASSERT(shard_cnt == psize + dsize);
+
+	shard = start_shard + dsize;
+	D_RWLOCK_RDLOCK(&obj->cob_lock);
+	if (obj->cob_version != map_ver) {
+		D_DEBUG(DB_IO, "ol ver %d != map ver %d\n", obj->cob_version, map_ver);
+		D_RWLOCK_UNLOCK(&obj->cob_lock);
+		return -DER_STALE;
+	}
+	for (i = 0; i < psize; i++, shard++) {
+		if (obj->cob_shards->do_shards[shard].do_rebuilding)
+			continue;
+		if (obj->cob_shards->do_shards[shard].do_target_id != -1)
+			break;
+	}
+	D_RWLOCK_UNLOCK(&obj->cob_lock);
+
+	/* check exist from one parity shard */
+	if (i < psize && likely(!DAOS_FAIL_CHECK(DAOS_OBJ_SKIP_PARITY))) {
+		D_DEBUG(DB_IO, DF_OID" choose parity shard %d\n",
+			DP_OID(obj->cob_md.omd_id), shard);
+		tgt = req_tgts->ort_shard_tgts;
+		/* obj_shard_tgts_query() should not goes to degraded fetch as long as the
+		 * map_ver not changed.
+		 */
+		rc = obj_shard_tgts_query(obj, map_ver, shard, shard - start_shard, tgt, obj_auxi);
+		if (rc == 0) {
+			req_tgts->ort_grp_nr = 1;
+			req_tgts->ort_grp_size = 1;
+		} else {
+			D_ERROR(DF_OID" obj_shard_tgts_query failed "DF_RC"\n",
+				DP_OID(obj->cob_md.omd_id), DP_RC(rc));
+		}
+		return rc;
+	}
+
+	/* check exist from all data shards */
+	shard = start_shard;
+	D_RWLOCK_RDLOCK(&obj->cob_lock);
+	if (obj->cob_version != map_ver) {
+		D_DEBUG(DB_IO, "ol ver %d != map ver %d\n", obj->cob_version, map_ver);
+		D_RWLOCK_UNLOCK(&obj->cob_lock);
+		return -DER_STALE;
```

Comment:
version was already compared above?

Reply:
yes, but seems possible that this thread being suspend by OS and resume executing,  in the window another thread refresh the layout.